### PR TITLE
[RLlib] Learner API: Fix and unify grad-clipping configs and behaviors.

### PR DESCRIPTION
**Pull request by sven1977**
Learner API: Fix and unify grad-clipping configs and behaviors.
* Introduce new AlgorithmConfig setting: `grad_clip_by`, which can be set to "value", "norm" or "global_norm" and determines the mode of clipping.
* Made `grad_clip` a generic AlgorithmConfig property (was only supported by some algos before). However, this setting is only used if `_enable_learner_api=True`.
* Implement proper clipping behaviors based on the 3 new modes in Tf- and TorchLearner (postprocess_gradients method).
* Add test cases for these new behaviors.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/learner/tf/tf_learner.py [hunk 1]
```diff
@@ -96,12 +97,49 @@ def compute_gradients(
...
+        grad_clip_by_value = self._optimizer_config.get("grad_clip_by_value", None)
+        if grad_clip_by_value is not None:
+            gradients_dict = {
+                k: tf.clip_by_value(
+                    v, -grad_clip_by_value, grad_clip_by_value
+                ) for k, v in gradients_dict.items()
+            }
+
+        # Clip by L2-norm (per gradient tensor).
+        grad_clip_by_norm = self._optimizer_config.get("grad_clip_by_norm", None)
```
Comment (avnishn) [inline rllib/core/learner/tf/tf_learner.py:117]: are we going to allow users to clip gradients by value and then by norm? that doesn't sound right to me, but that is the behavior that has been enabled here.

File: rllib/algorithms/qmix/qmix.py [hunk 2]
```diff
@@ -77,7 +77,10 @@ def __init__(self):
         self.double_q = True
         self.optim_alpha = 0.99
         self.optim_eps = 0.00001
-        self.grad_clip = 10
+
+        # TODO (sven): Deprecate grad_clip setting once all-in on new Learner API.
+        self.grad_clip = 10.0
+        self.grad_clip_by_global_norm = 10.0
```
Comment (avnishn) [inline rllib/algorithms/qmix/qmix.py:83]: how is this going to be used downstream by the qmix policy / multi_gpu_train_one_batch.

**Review by avnishn — COMMENTED** [review]
Thanks for sharing this. However, I am overall confused by this pr. Is it incomplete?

File: rllib/core/learner/tf/tf_learner.py [hunk 3]
```diff
@@ -96,12 +97,49 @@ def compute_gradients(
...
+        # Clip by value (each gradient individually).
+        grad_clip_by_value = self._optimizer_config.get("grad_clip_by_value", None)
+        if grad_clip_by_value is not None:
+            gradients_dict = {
+                k: tf.clip_by_value(
+                    v, -grad_clip_by_value, grad_clip_by_value
+                ) for k, v in gradients_dict.items()
+            }
+
+        # Clip by L2-norm (per gradient tensor).
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/core/learner/tf/tf_learner.py:115]: Great point. Was thinking about this myself. Maybe we should just do: `grad_clip: [some value]` and then `grad_clip_by: [value|norm|global_norm]`.

**Comment by sven1977** [comment]
Hey @avnishn , sorry about the confusion and thanks for taking a look!

I went through the PR once more and your comments and addressed all of them.
I changed the configs to the backward-compatible pair of `grad_clip` and (new) `grad_clip_by`, which specifies the mode of clipping. This way, it's a) backward compatible and b) exclusive (only one way of clipping at the same time allowed).

File: rllib/algorithms/impala/impala.py [hunk 4]
```diff
@@ -422,16 +425,6 @@ def validate(self) -> None:
             self.vtrace_clip_pg_rho_threshold
         )
 
-    @override(AlgorithmConfig)
```
Comment (sven1977) [inline rllib/algorithms/impala/impala.py:425]: Not needed anymore with this PR. grad clipping has been universally moved into `Learner.postprocess_gradients()`.

File: rllib/core/learner/torch/torch_learner.py [hunk 5]
```diff
@@ -79,6 +79,22 @@ def compute_gradients(
 
         return grads
 
+    @override(Learner)
```
Comment (sven1977) [inline rllib/core/learner/torch/torch_learner.py:82]: This was missing in torch thus far.

File: rllib/utils/tf_utils.py [hunk 6]
```diff
@@ -27,6 +27,48 @@
 tf1, tf, tfv = try_import_tf()
 
 
+@PublicAPI
```
Comment (sven1977) [inline rllib/utils/tf_utils.py:30]: New grad clip utilities, replacing the old, messy ones (some of the old ones do clipping by value, some by norm (not global norm), some have additional optimizer update logic in them, etc..

File: rllib/algorithms/qmix/qmix.py (hunk 2 above)
Comment (sven1977) [reply rllib/algorithms/qmix/qmix.py]: fixed by replacing the new settings with backward-compatible ones

**Review by avnishn — APPROVED** [review]

**Pull request merged**